[FLINK-18990][task] Read channel state sequentially

rkhachatryan opened the pull request:
## What is the purpose of the change

Read channel state sequentially to reduce the number of seeks.

## Verifying this change

- Added `SequentialChannelStateReaderImplTest` and `ChannelStateChunkReaderTest` for the new code.
- Added `RecoveredInputChannelTest` to check that conversion is not permitted untill fully recovered and state consumed
- some tests are eliminated because: 1) state in `StreamTask` is read unconditionally; 2) recovered channel can't be converted without being fully consumed anymore

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -504,23 +502,18 @@ protected void beforeInvoke() throws Exception {
-		} else {
-			requestPartitions();
+		});
+		for (InputGate inputGate : getEnvironment().getAllInputGates()) {
```
Okay you just did what I wrote on the last commit. Probably squashing makes sense.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/reader/AbstractRecordReader.java:
```diff
@@ -45,6 +45,8 @@
 
 	private RecordDeserializer<T> currentRecordDeserializer;
 
+	private boolean finishedStateReading;
```
nit: typo in commit message.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RecoveredInputChannel.java:
```diff
@@ -77,13 +78,18 @@
 	}
 
-	public abstract InputChannel toInputChannel() throws IOException;
+	public abstract InputChannel toInputChannelInternal() throws IOException;
```
`protected`?

rkhachatryan replied:
Yes :)

AHeise commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateSerializerImplTest.java:
```diff
@@ -36,7 +36,6 @@
 import java.util.Random;
 
 import static org.apache.flink.runtime.checkpoint.channel.ChannelStateByteBuffer.wrap;
```
Okay, now I see that you removed the original `ChannelStateReader`. I'd probably state that in the commit that adds `SeqReader`.

rkhachatryan replied:
Done.

AHeise commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateSerializerImplTest.java:
```diff
@@ -146,4 +145,10 @@ private void readAndCheck(byte[] data, ChannelStateSerializerImpl serializer, By
 		}
 	}
 
+	static byte[] generateData(int len) {
```
What's that change about?

rkhachatryan replied:
Previously, this method was imported from `ChannelStateReaderImplTest`.
In this commit `ChannelStateReaderImplTest` was removed and `generateData` moved to this class.

AHeise commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/ChannelPersistenceITCase.java:
```diff
@@ -76,17 +67,17 @@ public void testReadWritten() throws Exception {
 			singletonMap(resultSubpartitionInfo, resultSubpartitionInfoData)
 		);
 
-		assertArrayEquals(inputChannelInfoData, read(
```
It looks like some cleanup is needed here.

rkhachatryan replied:
Right, forgot about this test :)

AHeise reviewed (commented):
Very intriguing solution. I left a couple of comments but nothing deep. While reviewing my main issue was to understand the relationship between Sequential and old ChannelStateReader (become only clear after last commit). Last commits looks a bit incomplete.
Note that I still struggle a bit on the concepts of StateHandle and delegate. I dropped a question at the main commit.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -502,33 +503,24 @@ protected void beforeInvoke() throws Exception {
 					.thenRun(() -> mainMailboxExecutor.execute(inputGate::requestPartitions, "Input gate request partitions"));
 			}
+		} else {
+			requestPartitions();
```
Yes, this is what I'm doing in `[FLINK-18989][task] Read channel state unconditionally` (but I'm trying to move this commit to #13467).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java:
```diff
@@ -0,0 +1,263 @@
+	private <Info, Context, Handle extends AbstractChannelStateHandle<Info>> void read(
+			Function<OperatorSubtaskState, StateObjectCollection<Handle>> stateHandleExtractor,
+			RecoveredChannelStateHandler<Info, Context> stateHandler) throws IOException {
+		for (Map.Entry<StreamStateHandle, List<Handle>> delegateAndHandles : groupByDelegate(streamSubtaskStates(), stateHandleExtractor).entrySet()) {
```
> What is the relation between delegate and handles

1:M, that is one underlying delegate can be referenced by 1 or more handles.

> we only have one state file per subtask 

Yes, but on downscaling we can have more. 
Does this answer your question?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java:
```diff
@@ -0,0 +1,263 @@
+			source.seek(sourceOffset);
+		}
+		int length = serializer.readLength(source);
+		while (length > 0) {
```
length is decremented below: ```length -= serializer.readData```

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java:
```diff
@@ -0,0 +1,263 @@
+	}
+}
+
+class ChannelStateChunkReader {
```
Yes, mostly for testing.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java:
```diff
@@ -0,0 +1,263 @@
+		} catch (InterruptedException e) {
+			throw new IOException(e);
+		}
+		return Tuple2.of(wrap(bufferBuilder), Tuple2.of(bufferBuilder, bufferBuilder.createBufferConsumer()));
```
Here, it's important to create consumer before writing data.
From `BufferBuilder.createBufferConsumer` javadoc:
>  Data written to BufferBuilder before creation of BufferConsumer won't be visible for that BufferConsumer

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java:
```diff
@@ -0,0 +1,263 @@
+					length -= serializer.readData(source, bufferWithContext.f0, length);
+				}
+			} catch (Exception e) {
+				bufferWithContext.f0.recycle();
```
It's enough to clean buffer, context is not supposed to hold any resources on its own.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java:
```diff
@@ -0,0 +1,263 @@
+		try {
+			bufferBuilder = getWriter(subpartitionInfo).getBufferBuilder(subpartitionInfo.getSubPartitionIdx());
+		} catch (InterruptedException e) {
+			throw new IOException(e);
```
`InterruptedException` is a checked exception so I'd had to declare it if not wrapping. 
But it's not appropriate for this layer (IO).

`Thread.currentThread().interrupt();` only sets the status, so the caller can skip interrupt status check.

pnowojski replied:
`InterruptedException` quite often should be handled differently and it's good to know that the `InterruptedException` has happened.

For example, when caller received `IOException`, it might decide to retry or just keep going if that was not a critical part. `InterruptedException` should on the other hand shut down the code semi gracefully and as quickly as possible.

Here I think adding `throws InterruptedException` would be much better. But I didn't look how deep in a call stack are we and how difficult would it be to propagate it.

re `Thread.currentThread().interrupt();`, the crowd wisdom is to always set it when wrapping/hiding/rethrowing `InterruptedException`. I think we are using interrupted flag in only a couple of places, but it's better to keep it clean, just in case we would had to use it some important place and then we would discover we need to fix `interrupted` flag everywhere to make it work.

rkhachatryan replied:
You're right about `currentThread().interrupt()` and that handling of `IOException` likely differs from that of `InterruptedException`. 
But for the latter, I think it's better to wrap it into `RuntimeException`.

pnowojski replied:
Personally I don't like declared exceptions, but even in that case special treatment of the `InterruptedException` means  it makes sense to declare it explicitly. But the issue is that in Flink trend is to have explicit exceptions handling/declarations. Here it looks like the `InterruptedException` is already present on the call site (`StreamTask#readRecoveredChannelState`), so it shouldn't be an issue to conform with this.

rkhachatryan replied:
Removed `catch InterruptedException` and added to the method signature after not converging in an offline discussion.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReader.java:
```diff
@@ -0,0 +1,62 @@
+ * Reads channel state saved during checkpoint/savepoint.
+ */
+@Internal
+public interface SequentialChannelStateReader extends AutoCloseable {
```
Became apparent with later commits that it's going to be removed.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java:
```diff
@@ -0,0 +1,263 @@
+
+}
+
+interface RecoveredChannelStateHandler<Info, Context> extends AutoCloseable {
```
Good point, I'll move them to a separate file.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java:
```diff
@@ -0,0 +1,263 @@
+		this.chunkReader = new ChannelStateChunkReader(serializer);
+	}
+
+	public SequentialChannelStateReaderImpl(TaskStateSnapshot taskStateSnapshot, ChannelStateSerializer serializer, ChannelStateChunkReader chunkReader) {
```
I found this constructor can even be inlined.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java:
```diff
@@ -0,0 +1,263 @@
+	}
+
+	private static <Info, Handle extends AbstractChannelStateHandle<Info>> List<Tuple2<Long, Info>> extractOffsetsSorted(List<Handle> channelStateHandles) {
+		return channelStateHandles
```
Offsets are sorted only across single ChannelStateHandle. 
They are mostly sorted across all ChannelStateHandles referring to the same file (if handles are sorted in the same way as during writing).
So sorting is necessary here.

rkhachatryan commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -0,0 +1,162 @@
+
+		try (FSDataInputStream stream = geStream(serializer, 10)) {
+			new ChannelStateChunkReader(serializer).readChunk(stream, serializer.getHeaderLength(), handler, "channelInfo");
+			checkState(serializer.failed);
```
Yes, these `checkState`s should be in `finally`. Thanks!

rkhachatryan commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -0,0 +1,162 @@
+		@Override
+		public void close() throws Exception {
+		}
+
```
Do you mean double newline between two `}`? 
Why is it a problem?

It's very convenient for me to select it using vim select block :)

AHeise replied:
When looking at formatting, I usually just look for consistency and there are a few classes with this extra new line and many without. But which way is better is probably not worth debating ;).

rkhachatryan replied:
Removed newline.

rkhachatryan commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImplTest.java:
```diff
@@ -0,0 +1,321 @@
+	}
+
+	private final ChannelStateSerializer serializer;
+	private final Random random;
```
I think this wouldn't give performance benefit as tests are currently run sequentially.
And even if run in parallel, a new class instance would be created for each test, so threads wouldn't share `Random` instance.

AHeise replied:
The issue is rather that `Random` is always strictly slower than `ThreadLocalRandom` because of lock acquisitions. However, you are right, as long as we not using multiple threads, we do not have contention, so the difference is marginal.

rkhachatryan replied:
> Random is always strictly slower than ThreadLocalRandom

`nextBytes`  used in this test is not overridden in `ThreadLocalRandom` so it's the same.
And the creation of `Random` seems cheaper than `ThreadLocalRandom`.
So it would be slower in the end.

rkhachatryan commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImplTest.java:
```diff
@@ -0,0 +1,321 @@
+		return range(0, stateParLevel).boxed().flatMap(
+			gateId -> range(0, stateParLevel).mapToObj(
+				channelId ->
+					descriptorCreator.apply(gateId, channelId))).collect(toMap(
```
Good idea.
I'll extract `generateSingleChannelState` method.

rkhachatryan commented:
Thanks for reviewing @AHeise and @pnowojski 
I've addressed your feedback, please take a look.

rkhachatryan commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java:
```diff
@@ -197,22 +198,21 @@ public void testInputStarvation() throws Exception {
-			testHarness.processSingleStep();
 			expectedOutput.add(new StreamRecord<>("[2]: 2"));
-			assertThat(testHarness.getOutput(), contains(expectedOutput.toArray()));
+			testHarness.processAll();
+			assertEquals(expectedOutput, new ArrayList<>(testHarness.getOutput()).subList(0, expectedOutput.size()));
```
@pnowojski, 
this test started to fail after rebasing of `Read channel state unconditionally`.
So I replaced `processSingleStep` with `processAll`.
Please take a look.

pnowojski replied:
This test is now not doing what it was intended.

Now you are processing all elements from the input gate `1` before `testHarness.processElement(new StreamRecord<>("1"), 2);` (L207/206) is being enqueued to input gate `2`.

I would guess that 
```
// to avoid starvation, if the input selection is ALL and availableInputsMask is not ALL,
// always try to check and set the availability of another input
if (inputSelectionHandler.shouldSetAvailableForAnotherInput()) {
	fullCheckAndSetAvailable();
}
```
check from `StreamMultipleInputProcessor#selectNextReadingInputIndex` is currently not tested.

The intention behind this test is:
1. to have a long (just as well could be infinite) backlog of records to process on one of the inputs
2. introduce the availability change on the second input, and make sure it's checked/respected (instead of hot looping on the first input)
3. also throw in a third not selected input just to spice things a little bit

Why did you have to change this test?

rkhachatryan replied:
Thanks for the explanation.
The problem was that the output didn't match the expected one. The reason was that requesting partitions now also requires a mailbox step. 
After your comment I replaced the fix with:
```
for (int i = 0; i < testHarness.inputGates.length; i++) {
	testHarness.processSingleStep();
}
```
Does it look good to you?

pnowojski replied:
Hmmm, maybe replace it with more reliable condition, sth like:
```
boolean noElementFromInputGate3 = true;
steps = 0;
while (noElementFromInputGate3 && steps++ < 100 && testHarness.procesSingleStep()) {
  noElementFromInputGate3 = ... // check for presence of `new StreamRecord<>("[3]: 1"))`
}
```
?

AHeise commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/util/StreamTaskUtil.java:
```diff
@@ -29,9 +30,10 @@
+	public static void waitTaskIsRunning(StreamTask<?, ?> task, CompletableFuture<Void> taskInvocation) throws InterruptedException, ExecutionException {
 		while (!task.isRunning()) {
 			if (taskInvocation.isDone()) {
+				taskInvocation.get();
```
An alternative way to propagate the exception would be:
```
				taskInvocation.exceptionally(e -> {
					throw new AssertionError("Task has stopped", e);
				});
```

It depends if you want the test to fail through assertions or through unexpected exception.

rkhachatryan replied:
I'd prefer not to wrap it as it's more straightforward to me both in code and in the logs.

AHeise reviewed (commented):
Addressed most of my comments, a few open discussions.

I'm thinking that the last 3 commits (remove tests, methods, class of channel state reader) should be squashed.

rkhachatryan commented:
@flinkbot run azure

AHeise reviewed (approved):
LGTM % squash last 3 commits. I'm still waiting for @pnowojski to look at the changed test.

rkhachatryan commented:
Thanks for reviewing @AHeise and @pnowojski .
I've updated `StreamTaskMultipleInputSelectiveReadingTest.testInputStarvation` and squashed the last three commits.

pnowojski commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java:
```diff
@@ -183,6 +183,9 @@ public void testInputStarvation() throws Exception {
 					.setupOutputForSingletonOperatorChain(new TestInputStarvationMultipleInputOperatorFactory())
 					.build()) {
 
+			for (int i = 0; i < testHarness.inputGates.length; i++) {
+				testHarness.processSingleStep();
+			}
```
Would it work if replaced with `testHarness.processAll` ?

rkhachatryan commented:
Build failure unrelated: FLINK-19427
Private build passed.

rkhachatryan commented:
I rebased the PR and resolved conflicts with FLINK-19027 by moving the assignment of exclusive buffers for `RecoveredInputChannel` into it's `requestBufferBlocking` method.

I also replaced `RemoteInputChannel.assignExclusiveSegments()` with `InputChannel.setup()` make it clear that `RecoveredInputChannel` doesn't need setup and avoid if/casts (e4946c0fbc1a46d866b29edff083eb4e2370e437).

@pnowojski, @AHeise please take a look.

AHeise commented on flink-quickstart/flink-quickstart-java/src/main/resources/archetype-resources/${project.build.directory}/classes/log4j2.properties:
```diff
@@ -0,0 +1,25 @@
+# limitations under the License.
+################################################################################
+
+rootLogger.level = INFO
```
Something is wrong in this commit. I'm suspecting that you hit some weird maven/intellij interaction. Could you please delete all log4js properties?

rkhachatryan replied:
Yes, Intellij added these files and I accidentally committed.
Fixed.
